internal/backends/dart/dart.go

```diff
@@ -0,0 +1,250 @@
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"runtime"
+
+	"github.com/replit/upm/internal/api"
+	"github.com/replit/upm/internal/util"
+	"gopkg.in/yaml.v2"
+)
+
+// TODO: Properly implement package dir
+// Blocked on https://github.com/dart-lang/pub/issues/2009
+// Workaround inspired by https://github.com/google/pub_cache/blob/master/lib/pub_cache.dart#L17
+func dartGetPackageDir() string {
+	cacheEnv := os.Getenv("PUB_CACHE")
+	if cacheEnv != "" {
+		return cacheEnv
+	}
+
+	if runtime.GOOS == "windows" {
+		return path.Join(os.Getenv("APPDATA"), "Pub", "Cache")
+	}
+
+	return path.Join(os.Getenv("HOME"), ".pub-cache")
+}
+
+// PubspecYaml represents deps in a pubspec.yaml file.
+type dartPubspecYaml struct {
+	Name            string                 `yaml:"name"`
+	Description     string                 `yaml:"description"`
+	Dependencies    map[string]interface{} `yaml:"dependencies"`
+	DevDependencies map[string]interface{} `yaml:"dev_dependencies"`
+}
+
+// dartListPubspecYaml lists all deps in a pubspec.yaml file
+func dartListPubspecYaml() map[api.PkgName]api.PkgSpec {
+
+	var specs dartPubspecYaml
+	specs = readSpecFile()
+
+	pkgs := map[api.PkgName]api.PkgSpec{}
+	for nameStr, specStr := range specs.Dependencies {
+		pkgs[api.PkgName(nameStr)] = api.PkgSpec(dartParseSpec(specStr))
+	}
+	for nameStr, specStr := range specs.DevDependencies {
+		pkgs[api.PkgName(nameStr)] = api.PkgSpec(dartParseSpec(specStr))
+	}
+	return pkgs
+}
+
+// ParseDartSpec parses a Dart version handling special
+// cases like Flutter SDK deps where the spec is a map.
+func dartParseSpec(spec interface{}) string {
+	version, ok := spec.(string)
+	// Handle cases like Flutter SDK deps where the spec is a map.
+	if !ok {
+		version = "Unknown"
+	}
+	return version
+}
+
+// PubspecLock represents deps in a pubspec.lock file.
+type dartPubspecLock struct {
+	Packages map[string]struct {
+		Dependency string `yaml:"dependency"`
+		Source     string `yaml:"source"`
+		Version    string `yaml:"version"`
+	} `yaml:"packages"`
+}
+
+// dartListPubspecLock lists all deps in a pubspec.lock file
+func dartListPubspecLock() map[api.PkgName]api.PkgVersion {
+	contentsB, err := ioutil.ReadFile("pubspec.lock")
+	if err != nil {
+		util.Die("pubspec.lock: %s", err)
+	}
+	var cfg dartPubspecLock
+	if err := yaml.Unmarshal(contentsB, &cfg); err != nil {
+		util.Die("pubspec.lock: %s", err)
+	}
+	pkgs := map[api.PkgName]api.PkgVersion{}
+	for nameStr, data := range cfg.Packages {
+		pkgs[api.PkgName(nameStr)] = api.PkgVersion(data.Version)
+	}
+	return pkgs
+}
+
+// pubDevSearchResults represents the data we get from Pub.dev when
+// doing a search.
+type pubDevSearchResults struct {
+	Packages []struct {
+		Name string `json:"package"`
+	} `json:"packages"`
+	NextURL string `json:"next"`
+}
+
+// dartSearch implements Search for Pub.dev.
+func dartSearch(query string) []api.PkgInfo {
+	endpoint := "https://pub.dev/api/search"
```

Comment:
Make a global const `pubBaseUrl`

```golang
import "os"

// Base URL for pub repository.
//
// Still using pub.dartlang.org (instead of pub.dev), because this is used by the pub client, and
// by implication necessarily stable.
const defaultPubBaseUrl = "https://pub.dartlang.org"

String getPubBaseUrl() {
  baseUrl := os.Getenv("PUB_HOSTED_URL")
  if baseUrl != "" {
    return baseUrl
  }
  return defaultPubBaseUrl
}

func dartSearch (query string) []api.PkgInfo {
    endpoint := fmt.Sprintf("%s/api/search", pubBaseUrl);

   // ...
}
```